src/EventGridExtension/OutputBinding/EventGridAsyncCollector.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.EventGrid;
+using Microsoft.Azure.EventGrid.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.EventGrid
+{
+    public sealed class EventGridAsyncCollector : IAsyncCollector<EventGridEvent>
+    {
+        // use IEventGridClient for mocking test
+        private readonly IEventGridClient _client;
+        private readonly IList<EventGridEvent> _eventsToSend = new List<EventGridEvent>();
+        private readonly string _topicHostname;
+
+        private ManualResetEventSlim _canAdd = new ManualResetEventSlim(true);
+
+        public EventGridAsyncCollector(IEventGridClient client, string topicEndpointUri)
+        {
+            _client = client;
+            _topicHostname = new Uri(topicEndpointUri).Host;
+        }
+
+        public Task AddAsync(EventGridEvent item, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            _canAdd.Wait(); // spinlock while flushing takes place; avoid modifying collection during the Publish() operation
+
+            _eventsToSend.Add(item);
+
+            return Task.CompletedTask;
+        }
+
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                _canAdd.Reset();
+
+                if (_eventsToSend.Any())
```

Comment:
to optimize this, simply replace the collection within the scope of the lock and release the lock after.

Reply:
Not sure I follow.

Reply:
are we ok with creating a copy of the collection, using extra resources, to avoid using a mutex?

Reply:
You don't have to copy the collection. Replacing it with a new, empty collection is another option. Limiting the scope lock scope here would significantly help avoid contention, as the call to `PublishEventsAsync` is a much longer operation, and holding onto the lock for that duration of that call would prevent calls to `AddAsync` from returning until that is done.